src/datatable/sphinxext/xfunction.py

```diff
@@ -88,78 +88,121 @@
+        Global configuration options.
 
-        self.config : sphinx.config.Config
-            Global configuration options.
+    self.content : StringList
+        The content of the directive, i.e. after the directive itself and
+        all its options. This is a list of lines.
 
-        self.content : StringList
-            The content of the directive, i.e. after the directive itself and
-            all its options. This is a list of lines.
+    self.content_offset : int
+        Index of the first line of `self.content` within the source file.
 
-        self.content_offset : int
-            Index of the first line of `self.content` within the source file.
+    self.env : sphinx.environment.BuildEnvironment
+        Stuff.
 
-        self.env : sphinx.environment.BuildEnvironment
-            Stuff.
+    self.name : str
+        The name of the directive, such as "xfunction" or "xdata".
 
-        self.name : str
-            The name of the directive, such as "xfunction" or "xdata".
+    self.options : Dict[str, str]
+        Key-value store of all options passed to the directive.
 
-        self.options : Dict[str, str]
-            Key-value store of all options passed to the directive.
 
+    Fields derived from config/arguments/options
+    --------------------------------------------
+    self.module_name : str
+        The name of the module being documented, in our case it's
+        "datatable" (from config variable `xf_module_name`).
 
-    Derived from config/arguments/options
-    -------------------------------------
-        self.module_name -- config variable xf_module_name
-        self.project_root -- config variable xf_project_root
-        self.permalink_fn -- config variable xf_permalink_fn
-        self.obj_name -- python name of the object being documented
-        self.qualifier -- object's qualifier, this string + obj_name give a
-                          fully-qualified object name.
-        self.src_file -- name of the file where the code is located
-        self.src_fnname -- name of the C++ function (possibly with a namespace)
-                           which corresponds to the object being documented
-        self.src_fnname2 -- name of the second C++ function (for setters)
-        self.doc_file -- name of the file where the docstring is located
-        self.doc_var -- name of the C++ variable containing the docstring
-        self.test_file -- name of the file where tests are located
-        self.setter -- name of the setter variable (for :xdata: directives)
+    self.project_root : pathlib.Path
+        Location of the project's root folder, relative to the folder
+        where the documentation is built. The paths specified by the
+        `:src:` and `:doc:` options must be relative to this root
+        directory (from config variable `xf_project_root`).
+
+    self.permalink_url0 : str
+        Pattern to be used for constructing URLs to a file. This string
```

Comment:
What does it mean `constructing URLs to a file`?

Reply:
Suppose you have a file `tests/test-docs.py`. But what you want is a URL such as https://github.com/h2oai/datatable/blob/master/tests/test-docs.py where this file can be publicly located.

Reply:
Then "constructing URLs for a file" may be? Because URL stands for Uniform Resource Locator. Also, then I'm not sure why it is URLs, and not just URL?